File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow, fast = head, head 
        cycleFound = False
        
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                cycleFound = True
                break
        
        if not cycleFound:
            return None
        
        slow = head
        while slow is not fast:
            slow = slow.next
            fast = fast.next
        
        return slow

File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    if(head == null)
        return null;
    let slow = head, fast = head;
    
    while(fast !== null && fast.next !== null){
        slow = slow.next;
        fast = fast.next.next;
        if(fast == slow)
            break;
    }
    if(fast == null || fast.next == null)
        return null; // no cycle

    slow = head;
    while(slow !== fast){
        slow = slow.next;
        fast = fast.next;
    }
    return slow;
};